File: examples/official-sdk/java/src/main/java/McpServerExample.java
import java.util.List;
import java.util.Map;

import io.modelcontextprotocol.json.McpJsonDefaults;
import io.modelcontextprotocol.server.McpServer;
import io.modelcontextprotocol.server.McpSyncServer;
import io.modelcontextprotocol.server.transport.StdioServerTransportProvider;
import io.modelcontextprotocol.spec.McpSchema;
import io.modelcontextprotocol.spec.McpSchema.CallToolResult;
import io.modelcontextprotocol.spec.McpSchema.ServerCapabilities;
import io.modelcontextprotocol.spec.McpSchema.TextContent;
import io.modelcontextprotocol.spec.McpSchema.Tool;

public class McpServerExample {

    public static void main(String[] args) {
        // 1. Create the stdio transport provider (reads stdin, writes stdout)
        var transport = new StdioServerTransportProvider(McpJsonDefaults.getMapper());

        // 2. Define the "hello" tool
        Tool helloTool = Tool.builder()
            .name("hello")
            .description("Say hello to someone")
            .inputSchema(new McpSchema.JsonSchema(
                "object",
                Map.of("name", Map.of("type", "string", "description", "The person's name")),
                List.of("name"),
                null, null, null
            ))
            .build();

        // 3. Define the "add" tool
        Tool addTool = Tool.builder()
            .name("add")
            .description("Add two numbers together")
            .inputSchema(new McpSchema.JsonSchema(
                "object",
                Map.of(
                    "a", Map.of("type", "number", "description", "First number"),
                    "b", Map.of("type", "number", "description", "Second number")
                ),
                List.of("a", "b"),
                null, null, null
            ))
            .build();

        // 4. Build the synchronous server with both tools
        McpSyncServer server = McpServer.sync(transport)
            .serverInfo("example-server", "1.0.0")
            .capabilities(ServerCapabilities.builder().tools(true).build())
            .toolCall(helloTool, (exchange, request) -> {
                String name = (String) request.arguments().get("name");
                return CallToolResult.builder()
                    .content(List.of(new TextContent("Hello, " + name + "!")))
                    .build();
            })
            .toolCall(addTool, (exchange, request) -> {
                double a = ((Number) request.arguments().get("a")).doubleValue();
                double b = ((Number) request.arguments().get("b")).doubleValue();
                String json = "{\"sum\":" + (a + b) + "}";
                return CallToolResult.builder()
                    .content(List.of(new TextContent(json)))
                    .build();
            })
            .build();

        // Server is now running and processing stdio messages.
        // It will block until the transport is closed (e.g. stdin EOF).
    }
}

File: examples/official-sdk/kotlin/src/main/kotlin/McpServerExample.kt
import io.ktor.utils.io.streams.asInput
import io.modelcontextprotocol.kotlin.sdk.server.Server
import io.modelcontextprotocol.kotlin.sdk.server.ServerOptions
import io.modelcontextprotocol.kotlin.sdk.server.StdioServerTransport
import io.modelcontextprotocol.kotlin.sdk.types.CallToolResult
import io.modelcontextprotocol.kotlin.sdk.types.Implementation
import io.modelcontextprotocol.kotlin.sdk.types.ServerCapabilities
import io.modelcontextprotocol.kotlin.sdk.types.TextContent
import io.modelcontextprotocol.kotlin.sdk.types.ToolSchema
import kotlinx.coroutines.Job
import kotlinx.coroutines.runBlocking
import kotlinx.io.asSink
import kotlinx.io.buffered
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import kotlinx.serialization.json.putJsonObject

fun main() {
    val server = Server(
        Implementation(name = "example-server", version = "1.0.0"),
        ServerOptions(
            capabilities = ServerCapabilities(
                tools = ServerCapabilities.Tools(listChanged = true),
            ),
        ),
    )

    // Register "hello" tool
    server.addTool(
        name = "hello",
        description = "Say hello to someone",
        inputSchema = ToolSchema(
            properties = buildJsonObject {
                putJsonObject("name") {
                    put("type", "string")
                    put("description", "The person's name")
                }
            },
            required = listOf("name"),
        ),
    ) { request ->
        val name = request.arguments?.get("name")?.jsonPrimitive?.content ?: "world"
        CallToolResult(content = listOf(TextContent("Hello, $name!")))
    }

    // Register "add" tool
    server.addTool(
        name = "add",
        description = "Add two numbers together",
        inputSchema = ToolSchema(
            properties = buildJsonObject {
                putJsonObject("a") {
                    put("type", "number")
                    put("description", "First number")
                }
                putJsonObject("b") {
                    put("type", "number")
                    put("description", "Second number")
                }
            },
            required = listOf("a", "b"),
        ),
    ) { request ->
        val a = request.arguments?.get("a")?.jsonPrimitive?.doubleOrNull ?: 0.0
        val b = request.arguments?.get("b")?.jsonPrimitive?.doubleOrNull ?: 0.0
        CallToolResult(content = listOf(TextContent("""{"sum":${a + b}}""")))
    }

    // Create stdio transport and run
    val transport = StdioServerTransport(
        System.`in`.asInput(),
        System.out.asSink().buffered(),
    )

    runBlocking {
        val session = server.createSession(transport)
        val done = Job()
        session.onClose { done.complete() }
        done.join()
    }
}
